pkg/cmd/experimental/syncgroups/groupsyncer.go

```diff
@@ -0,0 +1,260 @@
+package syncgroups
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/go-ldap/ldap"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kerrs "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+
+	"github.com/openshift/origin/pkg/client"
+	syncconfig "github.com/openshift/origin/pkg/cmd/experimental/syncgroups/config"
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/schema1"
+	ouserapi "github.com/openshift/origin/pkg/user/api"
+)
+
+// GroupSyncer runs a Sync job on Groups
+type GroupSyncer interface {
+	Sync() (err error)
+}
+
+// LDAPGroupSyncer sync Groups with records on an external LDAP server
+type LDAPGroupSyncer struct {
+	// Lists all groups to be synced
+	groupLister interfaces.LDAPGroupLister
+	// Fetches a group and extracts object metainformation and membership list from a group
+	groupMemberExtractor interfaces.LDAPGroupMemberExtractor
+	// Maps an LDAP user entry to an OpenShift User's Name
+	userNameMapper interfaces.LDAPUserNameMapper
+	// Maps an LDAP group enrty to an OpenShift Group's Name
+	groupNameMapper interfaces.LDAPGroupNameMapper
+	// Allows the Syncer to search for OpenShift Groups
+	groupClient client.GroupInterface
+	// config stores all sync configuration
+	config syncconfig.LDAPSyncConfig
+}
+
+// Sync allows the LDAPGroupSyncer to be a GroupSyncer
+func (s *LDAPGroupSyncer) Sync() error {
+	// determine what to sync
+	ldapGroupUIDs, err := s.groupLister.ListGroups()
+	if err != nil {
+		return err
+	}
+
+	for _, ldapGroupUID := range ldapGroupUIDs {
+		// get membership data
+		memberEntries, err := s.groupMemberExtractor.ExtractMembers(ldapGroupUID)
+		if err != nil {
+			return err
```

Comment:
Rather than dying on a failure, make an error type that includes the `ldapGroupUID` and the error to return a slice of failures.